Fix pull_image and push_image to handle tag@digest format correctly

Laptop765 opened the pull request:
##### SUMMARY

This PR fixes a bug that caused `pull_image` and `push_image` to fail with "invalid tag format" errors when using Docker image references containing both a tag AND a digest (e.g., `nginx:1.21@sha256:abc123`).

**The Problem:**

The Docker SDK and API don't accept combined `tag@digest` in the tag parameter. When `parse_repository_tag()` returns `("nginx", "1.21@sha256:abc123")`, passing this directly to `pull()` or push API fails:

```
400 Client Error: Bad Request ("invalid tag format")
```

**The Solution:**

Added a `build_pull_arguments()` helper function that detects combined `tag@digest` format and restructures the parameters:

```python
def build_pull_arguments(name: str, tag: str) -> tuple[str, str | None]:
    """
    Build correct arguments for Docker pull/push operations.

    - Tag-only: "v1.0" -> (name, "v1.0")
    - Digest-only: "sha256:abc..." -> (name, "sha256:abc...")
    - Combined: "v1.0@sha256:abc..." -> ("name:v1.0@sha256:abc...", None)
    """
    if "@" in tag and not tag.startswith("sha256:"):
        tag_part, digest = tag.split("@", 1)
        return f"{name}:{tag_part}@{digest}", None
    return name, tag
```

Fixes #1207

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
- `build_pull_arguments` in `plugins/module_utils/_util.py` (new)
- `pull_image` in `plugins/module_utils/_common.py`
- `pull_image` in `plugins/module_utils/_common_api.py`
- `push_image` in `plugins/modules/docker_image.py`

##### ADDITIONAL INFORMATION

**Files Changed:**

1. `plugins/module_utils/_util.py` - Added `build_pull_arguments()` helper function
2. `plugins/module_utils/_common.py` - Updated SDK-based `pull_image()` to use helper
3. `plugins/module_utils/_common_api.py` - Updated API-based `pull_image()` to use helper
4. `plugins/modules/docker_image.py` - Updated `push_image()` to use helper
5. `tests/unit/plugins/module_utils/test_pull_image.py` - Comprehensive unit tests

**Test Coverage:**

18 unit tests covering:
- Tag-only formats (`nginx:1.21`, `ghcr.io/user/repo:v1.0`, `localhost:5000/app:v2.0`)
- Digest-only formats (`nginx@sha256:abc...`)
- Combined tag@digest formats (the bug case)
- Integration tests with `parse_repository_tag()` to verify end-to-end flow

**Example Fix (Pull):**

Before fix:
```python
# pull_image("portainer/portainer-ee", "2.35.0-alpine@sha256:abc...")
# -> Calls: pull("portainer/portainer-ee", tag="2.35.0-alpine@sha256:abc...")
# -> FAILS: "invalid tag format"
```

After fix:
```python
# pull_image("portainer/portainer-ee", "2.35.0-alpine@sha256:abc...")
# -> build_pull_arguments returns: ("portainer/portainer-ee:2.35.0-alpine@sha256:abc...", None)
# -> Calls: pull("portainer/portainer-ee:2.35.0-alpine@sha256:abc...")
# -> SUCCESS: Docker accepts full reference without tag parameter
```

**Example Fix (Push):**

Before fix:
```python
# push_image() calls /images/{name}/push?tag=2.35.0-alpine@sha256:abc...
# -> FAILS: "invalid tag format"
```

After fix:
```python
# push_image() calls /images/{name:tag@digest}/push (no tag param)
# -> SUCCESS: Docker accepts full reference in image name
```

All existing tests continue to pass, confirming backward compatibility.

Laptop765 committed:
Fix parse_repository_tag to handle images with both tag and digest

The parse_repository_tag() function was incorrectly parsing Docker image
references that contained both a tag and a digest (e.g., nginx:1.0@sha256:abc).

Previously, when splitting by '@' first, the tag would be included in the
repository name, resulting in incorrect parsing:
- Input: "nginx:1.0@sha256:abc123"
- Old output: ("nginx:1.0", "sha256:abc123")
- Expected: ("nginx", "1.0@sha256:abc123")

The fix now:
1. Checks for digest (@) separator first
2. Examines the part before the digest for a tag (:) separator
3. Combines tag and digest as "tag@digest" when both are present

Added test cases:
- test_index_image_tag_and_sha
- test_index_user_image_tag_and_sha
- test_private_reg_image_tag_and_sha

Laptop765 committed:
Fix _image_lookup to handle combined tag@digest format

The parse_repository_tag fix alone is not sufficient because Docker stores
RepoTags and RepoDigests separately. When looking up an image with combined
tag@digest (e.g., nginx:1.21@sha256:abc...), the _image_lookup function must
split the combined format and match BOTH RepoTags (for the tag) AND
RepoDigests (for the digest).

Docker stores:
- RepoTags: ["nginx:1.21"]
- RepoDigests: ["nginx@sha256:abc..."]

But NEVER stores the combined format. The previous code would construct:
- lookup = "nginx:1.21@sha256:abc..." (never matches RepoTags)
- lookup_digest = "nginx@1.21@sha256:abc..." (never matches RepoDigests)

This fix:
1. Adds filter_images_by_tag() helper function to _util.py to avoid code
   duplication between _common.py and _common_api.py
2. Detects combined tag@digest format in the tag parameter
3. Splits into tag_part and digest_part
4. Constructs proper lookups for both RepoTags and RepoDigests
5. Requires BOTH to match for successful image lookup

Without this fix, image_label_mismatch: ignore fails because the image
cannot be found, resulting in no image labels being included in expected
labels comparison.

Includes comprehensive unit tests in test__util.py covering all scenarios
including edge cases for multiple @ symbols and empty tag parts.

felixfontein commented on plugins/module_utils/_api/utils/utils.py:
```diff
@@ -240,9 +240,23 @@ def convert_service_networks(
 
 
 def parse_repository_tag(repo_name: str) -> tuple[str, str | None]:
```
This function is called in a lot of places, changing its semantics can introduce a lot of subtle bugs. I don't think it's a good idea to change the behavior of this function.

I think it's better to have proper parsing for images. I've started working on that a longer time ago, but then got interrupted and haven't continued it for a longer time. The original approach in Docker SDK for Python - where this code comes from - seems to be pretty broken, and causes a lot of problems in quite a few places in this collection (and likely in other users of that code).

felixfontein reviewed (changes requested):
Thanks for your contribution!

Laptop765 committed:
fix: Handle tag@digest format in pull_image methods

When using combined image:tag@digest references, parse_repository_tag
returns (repo, "tag@digest"). The Docker SDK and API don't accept
tag@digest in the tag parameter, causing "invalid tag format" errors.

This fix:
1. Adds build_pull_arguments() helper function to _util.py
2. Uses the helper in both pull_image implementations
3. When tag contains @ (but isn't a pure digest), passes the full
   reference as the repository/fromImage parameter instead of splitting

Tested formats:
- portainer/portainer-ee:2.35.0-alpine@sha256:abc...
- ghcr.io/gethomepage/homepage:v1.7@sha256:abc...
- localhost:5000/myapp:v2.0@sha256:abc...

The existing filter_images_by_tag already handles tag@digest for
lookups, so find_image continues to work correctly.

Includes comprehensive unit tests for build_pull_arguments().

Laptop765 committed:
fix: Handle tag@digest format in push_image method

Extends the tag@digest fix to also cover push operations in docker_image.py.
The push_image() method was passing combined tag@digest format directly to
the Docker API's /images/{name}/push endpoint, which fails with
"invalid tag format" errors.

This fix:
1. Imports build_pull_arguments() into docker_image.py
2. Uses the helper in push_image() before calling the API
3. When tag contains @ (but isn't a pure digest), passes the full
   reference as the image name and omits the tag parameter

This complements the previous fix to pull_image() methods, ensuring
both pull and push operations handle tag@digest correctly.

Laptop765 committed:
fix: Match by digest only for combined tag@digest image lookups

When using combined tag@digest references (e.g., nginx:1.21@sha256:abc...),
Docker does NOT store the tag in RepoTags. It only stores the digest in
RepoDigests. The previous implementation required BOTH to match, which
always failed because RepoTags was empty.

This caused docker_container to pull the image on every run even when
the image with the correct digest already existed locally, breaking
idempotency.

The fix: When a digest is specified, match by digest only since it's the
authoritative identifier. The tag is informational for human readability.

Real-world example from docker image inspect:
  "RepoTags": [],  # Empty when pulled by digest\!
  "RepoDigests": ["portainer/portainer-ee@sha256:7ecf2008..."]

Updated tests to reflect the correct behavior:
- test_empty_repo_tags_matches_by_digest (the critical fix case)
- test_combined_tag_digest_matches_even_if_tag_differs
- test_multiple_tags_irrelevant_for_combined
